vochain/state.go

```diff
@@ -2,28 +2,153 @@ package vochain
+		return nil, fmt.Errorf("cannot unmarshal StateDBProcess: %w", err)
+	}
+	return sdbProc.Process.CensusRoot, nil
+}
+
+// processSetCensusRoot is the SetRootFn function to set the census root of a
+// process leaf.
+func processSetCensusRoot(value []byte, root []byte) ([]byte, error) {
+	var sdbProc models.StateDBProcess
+	if err := proto.Unmarshal(value, &sdbProc); err != nil {
+		return nil, fmt.Errorf("cannot unmarshal StateDBProcess: %w", err)
+	}
+	sdbProc.Process.CensusRoot = root
+	newValue, err := proto.Marshal(&sdbProc)
+	if err != nil {
+		return nil, fmt.Errorf("cannot marshal StateDBProcess: %w", err)
+	}
+	return newValue, nil
+}
+
+// processGetVotesRoot is the GetRootFn function to get the votes root of a
+// process leaf.
+func processGetVotesRoot(value []byte) ([]byte, error) {
+	var sdbProc models.StateDBProcess
+	if err := proto.Unmarshal(value, &sdbProc); err != nil {
+		return nil, fmt.Errorf("cannot unmarshal StateDBProcess: %w", err)
+	}
+	return sdbProc.VotesRoot, nil
+}
+
+// processSetVotesRoot is the SetRootFn function to set the votes root of a
+// process leaf.
+func processSetVotesRoot(value []byte, root []byte) ([]byte, error) {
+	var sdbProc models.StateDBProcess
+	if err := proto.Unmarshal(value, &sdbProc); err != nil {
+		return nil, fmt.Errorf("cannot unmarshal StateDBProcess: %w", err)
+	}
+	sdbProc.VotesRoot = root
+	newValue, err := proto.Marshal(&sdbProc)
+	if err != nil {
+		return nil, fmt.Errorf("cannot marshal StateDBProcess: %w", err)
+	}
+	return newValue, nil
+}
+
+// OraclesCfg is the Oracles subTree configuration.
+var OraclesCfg = statedb.NewTreeSingletonConfig(
+	arbo.HashFunctionSha256,
+	"oracs",
+	256,
+	rootLeafGetRoot,
+	rootLeafSetRoot,
+)
+
+// ValidatorsCfg is the Validators subTree configuration.
+var ValidatorsCfg = statedb.NewTreeSingletonConfig(
+	arbo.HashFunctionSha256,
+	"valids",
+	256,
+	rootLeafGetRoot,
+	rootLeafSetRoot,
+)
+
+// ProcessesCfg is the Processes subTree configuration.
+var ProcessesCfg = statedb.NewTreeSingletonConfig(
+	arbo.HashFunctionSha256,
+	"procs",
+	256,
+	rootLeafGetRoot,
+	rootLeafSetRoot,
+)
+
+// CensusCfg is the Census subTree (found under a Process leaf) configuration
+// for a process that supports non-anonymous voting with rolling census.
+var CensusCfg = statedb.NewTreeNonSingletonConfig(
+	arbo.HashFunctionSha256,
+	"cen",
+	256,
+	processGetCensusRoot,
+	processSetCensusRoot,
+)
+
+// CensusPoseidonCfg is the Census subTree (found under a Process leaf)
+// configuration when the process supports anonymous voting with rolling
+// census.  This Census subTree uses the SNARK friendly hash function Poseidon.
+var CensusPoseidonCfg = statedb.NewTreeNonSingletonConfig(
+	arbo.HashFunctionPoseidon,
+	"cenPos",
+	64,
+	processGetCensusRoot,
+	processSetCensusRoot,
+)
+
+// VotesCfg is the Votes subTree (found under a Process leaf) configuration.
+var VotesCfg = statedb.NewTreeNonSingletonConfig(
```

Comment:
I also wonder if using composite literals for these would be better; right now I have no idea what `256` means, but if you had `HashSize: 256` it would be a lot clearer. (I'm just making up that this means a hash size)

Reply:
If you need constructors, another way is to use named types for the parameters, so then you can end up with `statedb.HashSize(256)`.

Reply:
> I also wonder if using composite literals for these would be better; right now I have no idea what `256` means, but if you had `HashSize: 256` it would be a lot clearer. (I'm just making up that this means a hash size)

Yeah, this is something that bothered me a bit as well.  I wanted to use constructors so that the generated type can't be modified externally, but it was not clear what each parameter was.  I added a new struct type `TreeParams` to pass the parameters to the constructor and now it's much more clear!  thanks!

Now:
```go
	ValidatorsCfg = statedb.NewTreeSingletonConfig(statedb.TreeParams{
		HashFunc:          arbo.HashFunctionSha256,
		KindID:            "valids",
		MaxLevels:         256,
		ParentLeafGetRoot: rootLeafGetRoot,
		ParentLeafSetRoot: rootLeafSetRoot,
	})
```